### PR TITLE
Documentation listing supported layers for Keras, Caffe and TensorFlow

### PR DESCRIPTION
**Pull request by Azamlynny**

### CODE REVIEW

**Comment by thatbrguy** [comment]
@Azamlynny That's a very neat documentation, good job! I have a few suggestions before we can approve this:
- Place the documentation in `docs/source`. That's where most of our documentation is.
- Link this documentation to our README.md file.
- Please open a new PR with the graph reset functionality that you added for your previous task. Looks like you closed the old one and did not create another PR for graph reset.

File: tutorials/supported_layers.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+### Layer Wrappers
+| Layer                     | Caffe         | Keras        | Tensorflow  |
+| :-----------------------: | :-----------: | :----------: | :---------: |
+| TimeDistributed           | ×             | √            | √           |
+| Bidirectional             | ×             | √            | √           |
+
+### Custom Layers
+| Caffe         | Keras        | Tensorflow  |
+| :-----------: | :----------: | :---------: |
+| √             | √            | Use Keras API for custom layers |
```
Comment (Ram81) [inline tutorials/supported_layers.md]: You can add LRN as an example in custom layers supported.

File: tutorials/supported_layers.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+## Additional Notes:
+* Keras does not support the LRN layer used in Alexnet & many other models. To use the LRN layer refer to here: https://github.com/Cloud-CV/Fabrik/blob/master/tutorials/keras_custom_layer_usage.md.
+* Documentation for writing your own Keras layers is found here: https://keras.io/layers/writing-your-own-keras-layers/ 
+* Caffe 3D layers can be used in Caffe 2.0
```
Comment (Ram81) [inline tutorials/supported_layers.md]: We haven't tested any features for caffe 2.0 yet, have you tested this if not we can drop this line.

File: tutorials/supported_layers.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+## Documentation for Caffe, Keras, and Tensorflow layers
+* Documentation for all Keras Layers is found here: https://keras.io/layers/about-keras-layers/
```
Comment (Ram81) [inline tutorials/supported_layers.md]: You can add a markdown link on `here`.

File: tutorials/supported_layers.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+* Documentation for all Caffe Layers is found here: http://caffe.berkeleyvision.org/tutorial/layers.html
```
Comment (Ram81) [inline tutorials/supported_layers.md]: Same as above

File: tutorials/supported_layers.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+* Documentation for all Tensorflow Layers is found here: https://www.tensorflow.org/api_docs/python/tf/layers
```
Comment (Ram81) [inline tutorials/supported_layers.md]: Same as above

File: tutorials/supported_layers.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+| Bidirectional             | ×             | √            | √           |
+
+### Custom Layers
+| Caffe         | Keras        | Tensorflow  |
+| :-----------: | :----------: | :---------: |
+| √             | √            | Use Keras API for custom layers |
+
+## Additional Notes:
+* Keras does not support the LRN layer used in Alexnet & many other models. To use the LRN layer refer to here: https://github.com/Cloud-CV/Fabrik/blob/master/tutorials/keras_custom_layer_usage.md.
+* Documentation for writing your own Keras layers is found here: https://keras.io/layers/writing-your-own-keras-layers/
```
Comment (Ram81) [inline tutorials/supported_layers.md]: You can add a markdown link on here.

File: tutorials/supported_layers.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+| :-----------------------: | :-----------: | :----------: | :---------: |
+| TimeDistributed           | ×             | √            | √           |
+| Bidirectional             | ×             | √            | √           |
+
+### Custom Layers
+| Caffe         | Keras        | Tensorflow  |
+| :-----------: | :----------: | :---------: |
+| √             | √            | Use Keras API for custom layers |
+
+## Additional Notes:
```
Comment (Ram81) [inline tutorials/supported_layers.md]: You can add a markdown link on here.

File: tutorials/supported_layers.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
+# Layers supported in Caffe, Keras, and Tensorflow
+Below is are table showing which layers are supported by Caffe, Keras, and Tensorflow:
```
Comment (Ram81) [inline tutorials/supported_layers.md]: Grammatical mistake.

File: docs/source/supported_layers.md [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+| Cropping3D                | √             | √            | √           |
+| Upsampling 1D             | √             | √            | √           |
+| Upsampling 2D             | √             | √            | √           |
+| Upsampling 3D             | √             | √            | √           |
+| ZeroPadding 1D            | ×             | √            | √           |
+| ZeroPadding 2D            | ×             | √            | √           |
+| ZeroPadding 3D            | ×             | √            | √           |
+| Im2Col                    | √             | ×            | ×           |
+| Spatial Pyramid Pooling   | √             | ×            | ×           |
+* Upsampling in Caffe can be done by using methods shown here: https://gist.github.com/tnarihi/54744612d35776f53278
```
Comment (Ram81) [inline docs/source/supported_layers.md]: Add as a markdown link on `here`

File: docs/source/supported_layers.md [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+| AlphaDropout              | √             | √            | √           |
+ ### Layer Wrappers
+| Layer                     | Caffe         | Keras        | Tensorflow  |
+| :-----------------------: | :-----------: | :----------: | :---------: |
+| TimeDistributed           | ×             | √            | √           |
+| Bidirectional             | ×             | √            | √           |
+ ### Custom Layers
+| Layer         | Caffe        | Keras       | Tensorflow  |
+| :-----------: | :----------: | :---------: | :---------: |
+| Custom Layers | √            | √           | Use Keras API for custom layers |
```
Comment (Ram81) [inline docs/source/supported_layers.md]: No need to mention custom layers explicitly

**Comment by Ram81** [comment]
@Azamlynny minor changes

**Comment by yashdusing** [comment]
Also, could you please specify how you've tested the layers and the procedure used. We will need to cross verify it once on our end as well before approving the PR. A code template would be helpful as well .

**Comment by thatbrguy** [comment]
@Azamlynny Please add a section for `Data` layers. I've just pasted the layers here for your reference:

data

File: docs/source/supported_layers.md [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+| Stochastic Pooling        | √             | ×            | ×           | √             | ×               | ×                    |
+ 
+ ### Data Layers
+| Layer                     | Caffe         | Keras        | Tensorflow  | Fabrik Caffe  | Fabrik Keras    | Fabrik Tensorflow  |
+| :-----------------------: | :-----------: | :----------: | :---------: | :-----------: | :--------------: | :-----------------: | 
+| Image Data                | √             | √            | √           | √             | √               | √                    |
+| Data                      | √             | √            | √           | √             | √               | √                    |
+| HDF5 Data                 | √             | √            | √           | √             | √               | √                    |
+| HDF5 Output Data          | √             | √            | √           | √             | √               | √                    |
+| Input                     | √             | √            | √           | √             | √               | √                    |
```
Comment (thatbrguy) [inline docs/source/supported_layers.md]: If i remember correctly, for Keras and Tensorflow (in Fabrik) only `Input` is available. All other data layers are not available.

File: docs/source/supported_layers.md [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+| SELU                      | ×             | √            | √           | ×             | ×                | √                   |
+| Absolute Value            | √             | ×            | ×           | √             | ×                | ×                   |
+| Power                     | √             | √            | ×           | √             | ×                | ×                   |
+| Exp                       | √             | √            | ×           | √             | ×                | ×                   |
+| Linear                    | ×             | √            | √           | ×             | ×                | ×                   |
+| Log                       | √             | √            | ×           | √             | ×                | ×                   |
+| BNLL                      | √             | ×            | ×           | √             | ×                | ×                   |
+| Bias                      | √             | ×            | ×           | √             | √                | √                   |
+| Scale                     | √             | ×            | ×           | √             | √                | √                   |
```
Comment (thatbrguy) [inline docs/source/supported_layers.md]: I think Scale is Caffe only (in Fabrik)

File: docs/source/supported_layers.md [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+ ### Utility Layers
+| Layer                     | Caffe         | Keras        | Tensorflow  | Fabrik Caffe  | Fabrik Keras    | Fabrik Tensorflow  |
+| :-----------------------: | :-----------: | :----------: | :---------: | :-----------: | :--------------: | :-----------------: | 
+| Slicing                   | √             | ×            | ×           | √             | ×                | ×                   |
+| Eltwise                   | √             | ×            | ×           | √             | √                | √                   |
+| Parameter                 | √             | ×            | ×           | √             | ×                | √                   |
+| Reduction                 | √             | ×            | ×           | √             | ×                | ×                  |
+| Silence                   | √             | ×            | ×           | ×             | ×                | ×                   |
```
Comment (thatbrguy) [inline docs/source/supported_layers.md]: Fabrik Caffe does has silence.

File: docs/source/supported_layers.md [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+ ### Loss Layers
+| Layer                     | Caffe         | Keras        | Tensorflow  | Fabrik Caffe  | Fabrik Keras    | Fabrik Tensorflow  |
+| :-----------------------: | :-----------: | :----------: | :---------: | :-----------: | :--------------: | :-----------------: | 
+| Multinomial Logistic Loss | √             | ×            | ×           | ×             | ×                | ×                   |
```
Comment (thatbrguy) [inline docs/source/supported_layers.md]: Fabrik caffe has this layer

File: docs/source/supported_layers.md [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+| Infogain Loss             | √             | ×            | ×           | √             | ×                | ×                   |
+| Softmax with Loss         | √             | ×            | √           | √             | ×                | ×                   |
+| Sum-of-Squares/Euclidean  | √             | ×            | ×           | ×             | ×                | ×                   |
```
Comment (thatbrguy) [inline docs/source/supported_layers.md]: Fabrik caffe has this layer

File: docs/source/supported_layers.md [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+| Hinge / Margin            | √             | √            | √           | √             | ×                | ×                   |
+| Sigmoid Cross-Entropy Loss| √             | ×            | √           | ×             | ×                | ×                   |
```
Comment (thatbrguy) [inline docs/source/supported_layers.md]: Fabrik caffe has this layer

**Comment by yashdusing** [comment]
Hello @Azamlynny . Pretty good work so far. Few changes though (I've attached a screenshot completing the core layers) : 
- follow the format as show in picture i.e. if layer is keras only or caffe only then mention it in bracket. If layer is keras only, then even tensorflow would support it through keras API so mention keras in tensorflow column. If layer is supported in all of the frameworks (for eg. dropout), keep it as blank (i.e. don't add in brackets). Also, please refer the documentations for the frameworks before marking it as done. While caffe does have activation layers, it does not have a separate activation layer like keras does and has a separate layer for each of them which are mentioned layer (relu, sigmoid, tanH, etc) so activation layer should be crossed in caffe column.
- Reference : 
1. For caffe : http://caffe.berkeleyvision.org/tutorial/layers.html
2. For keras : https://keras.io/
3. For tensorflow 1.4 : https://github.com/tensorflow/docs/blob/r1.4/site/en/api_docs/api_docs/python/index.md
4. For Fabrik caffe : data.js
5. For Fabrik Keras : Fabrik's keras app's import, export files
5. For Fabrik Tensorflow : import_graphdef.py

**Comment by yashdusing** [comment]
@Ram81 @thatbrguy Please look into the format of the screenshot I posted and see if I missed something out or any changes are needed.

**Comment by thatbrguy** [comment]
@yashdusing I think we'll just mention the different names across platforms in brackets. (Inner-Product vs Dense). Let's not mention keras in the TensorFlow section for now. We'll have a separate task to deal with supported tensorflow APIs.

@Azamlynny Also, looks like inner-product is crossed out for Caffe. It's available on caffe so please modify that

**Comment by Azamlynny** [comment]
> @yashdusing I think we'll just mention the different names across platforms in brackets. (Inner-Product vs Dense). Let's not mention keras in the TensorFlow section for now. We'll have a separate task to deal with supported tensorflow APIs.

I am not sure by what you mean by mention the different names across platforms in brackets. Where would you want this? In the Layer column, or in the actual Caffe and Keras columns and if in these locations what would the layer column be called?

**Comment by thatbrguy** [comment]
@Azamlynny Actually the table looks neater with just ticks. So, if you notice any layer available in Caffe as well as Keras or Tensorflow, with just different names (but mapped to the same Fabrik layer name) you can add it to the notes section after the table

File: docs/source/supported_layers.md [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+| Input                     | √             | √            | √           | √             | √               | √                    |
+| Window Data               | √             | ×            | ×           | √             | ×               | ×                    |
+| Memory Data               | √             | √            | √           | √             | ×               | ×                    |
+| Dummy Data                | √             | ×            | ×           | √             | ×               | ×                    |
+| Python                    | √             | √            | √           | √             | ×               | ×                    |
+ 
+### Locally-connected Layers
+| Layer                     | Caffe         | Keras        | Tensorflow  | Fabrik Caffe  | Fabrik Keras    | Fabrik Tensorflow  |
+| :-----------------------: | :-----------: | :----------: | :---------: | :-----------: | :--------------: | :-----------------: | 
+| LocallyConnected1D        | ×             | √            | √           | ×             | √                | √×                  |
```
Comment (Ram81) [inline docs/source/supported_layers.md]: Why are there both `√× ` in the Fabrik Tensorflow section?

File: docs/source/supported_layers.md [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+| GlobalMaxPooling1D        | ×             | √            | √           | ×             | ×               | ×                    |
+| GlobalAveragePooling1D    | ×             | √            | √           | ×             | ×               | ×                    |
+| GlobalMaxPooling2D        | ×             | √            | √           | ×             | ×               | ×                    |
+| GlobalAveragePooling2D    | ×             | √            | √           | ×             | ×               | ×                    |
+| GlobalMaxPooling3D        | ×             | √            | √           | ×             | ×               | ×                    |
+| GlobalAveragePooling3D    | ×             | √            | √           | ×             | ×               | ×                    |
+| Stochastic Pooling        | √             | ×            | ×           | √             | ×               | ×                    |
+ 
+ ### Data Layers
+| Layer                     | Caffe         | Keras        | Tensorflow  | Fabrik Caffe  | Fabrik Keras    | Fabrik Tensorflow  |
```
Comment (Ram81) [inline docs/source/supported_layers.md]: Can you also tell me why we have 6 columns two for each layer?
I think only mentioning which layers for a framework does Fabrik support is good enough.
